### PR TITLE
Feature/uc3 (return RED when predicting signal outside available interval)

### PR DESCRIPTION
**Pull request by MishkaMN**
# PR Details
## Description

- Support dynamic SPAT when predicting signal states. Added start_time_stamps which is essential when it is dynamic.
- If time_stamp requested is outside available dynamic spat time interval, return STOP_AND_WAIT

## Related Issue
Part of assorted fixes here. Supports this PR:
https://github.com/usdot-fhwa-stol/carma-platform/pull/1919

## Motivation and Context
TSMO UC3.

## How Has This Been Tested?
Integration tested on Black Pacifica with UC3 MAP/SPAT

## Types of changes

- [X] Defect fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [X] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: common/lanelet2_extension/lib/CarmaTrafficSignal.cpp [hunk 1]
```diff
@@ -125,6 +125,37 @@ boost::optional<std::pair<boost::posix_time::ptime, CarmaTrafficSignalState>> Ca
...
   }
   
+  if (lanelet::time::toSec(fixed_cycle_duration) < 1.0) // there are recorded states, but no fixed_cycle_duration means it is dynamic
+  {
+    if (recorded_time_stamps.size() != recorded_start_time_stamps.size())
+    {
+      throw std::invalid_argument("recorded_start_time_stamps size is not equal to recorded_time_stamps size");
+    }
+    
+    // if requested time is BEFORE recorded states' time interval, return STOP_AND_WAIT
```
Comment (msmcconnell) [inline common/lanelet2_extension/lib/CarmaTrafficSignal.cpp]: typo STOP_AND_REMAIN

File: common/lanelet2_extension/lib/CarmaTrafficSignal.cpp [hunk 2]
```diff
@@ -125,6 +125,37 @@ boost::optional<std::pair<boost::posix_time::ptime, CarmaTrafficSignalState>> Ca
...
+    if (recorded_start_time_stamps.front() >= time_stamp)
+    {
+      return std::pair<boost::posix_time::ptime, CarmaTrafficSignalState>(recorded_start_time_stamps.front(), CarmaTrafficSignalState::STOP_AND_REMAIN);
+    }
+
+    // if requested time is AFTER recorded states' time interval, return STOP_AND_WAIT
```
Comment (msmcconnell) [inline common/lanelet2_extension/lib/CarmaTrafficSignal.cpp]: Typo STOP_AND_REMAIN

File: common/lanelet2_extension/lib/CarmaTrafficSignal.cpp [hunk 3]
```diff
@@ -125,6 +125,37 @@ boost::optional<std::pair<boost::posix_time::ptime, CarmaTrafficSignalState>> Ca
...
+    if (recorded_time_stamps.back().first <= time_stamp)
+    {
+      auto end_infinity_time = timeFromSec(2147483647);
+      LOG_WARN_STREAM("CarmaTrafficSignal doesn't have enough state saved, so returning STOP_AND_WAIT state! End_time: " << end_infinity_time);
```
Comment (msmcconnell) [inline common/lanelet2_extension/lib/CarmaTrafficSignal.cpp]: Typo STOP_AND_REMAIN

File: common/lanelet2_extension/lib/CarmaTrafficSignal.cpp [hunk 4]
```diff
@@ -125,6 +125,37 @@ boost::optional<std::pair<boost::posix_time::ptime, CarmaTrafficSignalState>> Ca
...
+      return std::pair<boost::posix_time::ptime, CarmaTrafficSignalState>(end_infinity_time, CarmaTrafficSignalState::STOP_AND_REMAIN); //end at 03:14:07 on Tuesday, 19 January 2038.
```
Comment (msmcconnell) [inline common/lanelet2_extension/lib/CarmaTrafficSignal.cpp]: Confusing comment //end at 03:14:07 on Tuesday, 19 January 2038.

**Review by msmcconnell — CHANGES_REQUESTED** [review]

File: common/lanelet2_extension/lib/CarmaTrafficSignal.cpp (hunk 4 above)
Comment (MishkaMN) [reply common/lanelet2_extension/lib/CarmaTrafficSignal.cpp]: I moved the comment to next to the actual number

**Review by msmcconnell — APPROVED** [review]

**Pull request merged**